File: class_primer/class_primer__closed_maze.cs
using System.Linq;
using static System.Console;
using Tpl = System.Tuple<char, int, int>;

internal class Program
{
    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        var l = new System.Collections.Generic.List<char>( n[1] + 1 );
        var a = new Tpl[n[0]];

        foreach( int i in Enumerable.Range( 0, n[0] ) ) {
            string[] t = ReadLine().Split();

            a[i] = new Tpl( t[0][0], int.Parse( t[1] ) - 1, int.Parse( t[2] ) - 1 );
        }
        l.Add( a[--n[2]].Item1 );
        foreach( int _ in Enumerable.Range( 0, n[1] ) ) {
            int m = int.Parse( ReadLine() );

            if( m == 1 ) {
                n[2] = a[n[2]].Item2;
            } else if( m == 2 ) {
                n[2] = a[n[2]].Item3;
            }
            l.Add( a[n[2]].Item1 );
        }
        WriteLine( string.Join( "", l ) );
        return;
    }
}

File: class_primer/class_primer__closed_maze.py
class maze:
    def __init__(self, a):
        self.a  = a[0]
        self.r1 = int(a[1]) - 1
        self.r2 = int(a[2]) - 1

n, k, s = list(map(int, input().split()))
l = []
o = []

for _ in range(n):
    l.append(maze(input().split()))
s -= 1
o.append(l[s].a)
for _ in range(k):
    s = l[s].r1 if input() == "1" else l[s].r2
    o.append(l[s].a)
print("".join(o))
